dm-constraints looked at me sideways, so I rewrote it.

emmanuel opened the pull request:
This is a partial re-write of `dm-constraints` (the non-adapter portion). I didn't redesign anything, just refactored and organized, and made it much more understandable.

emmanuel committed:
First step (a big one) towards a partial re-write.

emmanuel committed:
Clean up variable names, alignment, docs in adapter modules.

emmanuel committed:
Simplify JRuby check.

emmanuel committed:
Reformat lambda in spec.

emmanuel committed:
Update require spec to reflect normal use of Model.append_extensions.

emmanuel committed:
Move adapter support modules into their new home.

emmanuel committed:
ManyToOne inherits from OneToMany; make sure the latter is loaded first.

emmanuel committed:
Add an @todo to rename #auto_migrate_down_constraints!.

dkubb commented on lib/data_mapper/constraints/adapters/do_adapter.rb:
```diff
@@ -77,16 +84,18 @@ module DataMapper
-          source_model = relationship.source_model
-          source_table = source_model.storage_name(name)
+          storage_name = relationship.source_model.storage_name(name)
+          constraint_name     = constraint_name(storage_name, relationship.name)
```
This is a small nitpick, but generally when I have a block of assignments I vertically align the "=" signs. I find it more readable than "jagged" formatting.

dkubb commented on lib/data_mapper/constraints/migrations/model.rb:
```diff
@@ -0,0 +1,45 @@
+          auto_migrate_down_constraints!(repository_name)
+          super
+          auto_migrate_up_constraints!(repository_name)
+        end
```
I'm not sure why `auto_migrate_down_constraints!` and `auto_migrate_up_constraints!` have default arguments. Unless they're being called by other code, then it's likely `repository_name` will always be passed in.

I might also recommend sort of "namespacing" the method names so they follow the convention `auto_migrate_constraints_*` since they're being mixed into model. What would be ideal is if everything mixed into model prefixed the methods with something consistent that won't collide; they currently don't, but that doesn't mean we can't start with this.

I also wonder about that `is_remixable?` test. I don't know much about remixable, but I wonder what behaviour remixable models have that makes them not work with constraints. If we could figure that out we could establish some other convention to signal that a model shouldn't have constraints, rather than testing for a concern that is coupled to dm-is-remixable.

dkubb commented on lib/data_mapper/constraints/migrations/singleton_methods.rb:
```diff
@@ -0,0 +1,37 @@
+          repository_execute(:auto_migrate_down_constraints!, repository_name)
+          descendants = super
+          repository_execute(:auto_migrate_up_constraints!, repository_name)
+          descendants
```
@emmanuel: Do you know if it's part of the contract to return `descendants` here?

In general I try to closely follow [CQRS](http://en.wikipedia.org/wiki/Command-query_separation) when writing new code, or when changing old code with no well defined contracts. In this case that method is a "command" method, so (ideally) it should return `self`. A "query" method is idempotent given a specific object state, and while they are normally boolean methods, they could be anything. In fact, in veritas I go so far as to assert all query methods return an object that always returns true when compared using `#equal?` with an earlier value using from the same method -- I don't know if that's possible in DM code, but that's the general spirit I strive for.

dkubb commented on spec/isolated/require_spec.rb:
```diff
@@ -1,11 +1,14 @@
-    DataMapper::Model.respond_to?(:auto_migrate_down_constraints!, true).should be(true)
-    DataMapper::Model.respond_to?(:auto_migrate_up_constraints!,   true).should be(true)
+  it "extends Model descendants with the constraint API" do
+    DataMapper::Model.descendants.any?.should be(true)
```
You could probably write this more clearly as `DataMapper::Model.descendants.should_not be_empy`.

While `Enumerable#any?` and `Array#empty?` are not _exact_ inverses, in this case it should work fine.

dkubb commented:
@emmanuel except for a couple of small notes/questions this looks awesome to me.

emmanuel committed:
Align assignments.

emmanuel committed:
Correct documented type of arg.

emmanuel committed:
Remove dead comment.

emmanuel committed:
Integrate dkubb's remaining suggestions.

Better command/query separation (commands return self), pseudo-namespace methods piled into Model and Relationship. 

And one of my own, drop '!' on non-public auto_migrate methods, because there is no non-destructive alternative, that's just the behavior of the command.

emmanuel commented:
OK, I think I've covered your suggestions. Let me know if there's anything else I should clean up or change before this is good to go.

solnic commented:
Would be nice to squeeze those commits into one.

dkubb commented:
I think this is good. I don't think there's any specific need to collapse these into a single commit though. I have no hard/fast rule over when collapsation should happen, but I generally do it when adding a new feature. For refactoring I think it's ok to "show your work" and how you got to the solution through iterative improvement. I dunno, I guess I'd say use whichever you think is best in this case.

emmanuel commented:
I figure that's what merge commits are for: pseudo-collapsation from the perspective of the destination, without losing the information in the source. 

That said... I don't have commit bit here on `dm-constraints`, so I need someone else to make the final call. I'm fine either way.

dkubb commented:
@emmanuel you should have the commit bit now.

emmanuel committed:
Mark @api private methods as such.

solnic commented:
My point with merging into one is that it's easier to revert one commit in case something went wrong.

emmanuel commented:
Oh I see, I didn't think of that. 

Hmm, interesting; I've never tried to revert a merge commit, I wonder how tricky that is.

solnic commented:
@emmanuel Right, so it's _not_ as easy as git revert {sha}. Personally I'm a fan of squeezing commits from topic branches. The less commits we have the better, in general.
